### ISSUE TITLE
feature_copy = feature.clone().detach()

### ISSUE THREAD

**Issue by Walstruzz** [hyao1/ECC#1]
感谢您的工作, 我最近在我的工作里复现ECCLoss, 发现在细粒度小数据集上的确提升明显, 不过当我和这里的ECCLoss实现进行对比的时候, 我发现有一些细微的差异, 有一些疑惑的地方, 希望能帮忙解答

https://github.com/hyao1/ECC/blob/6a214246361b1b039ddcfe0041bc7e0117f80d4e/utils.py#L29
这里detach出来之后, `feature_copy`和`feature_table`都没计算梯度, 那`feature_intra_loss` 是否会回传梯度? 

https://github.com/hyao1/ECC/blob/6a214246361b1b039ddcfe0041bc7e0117f80d4e/utils.py#L43
这里为什么要做归一化, 我看论文中没有提及, 也可能是我遗漏了什么信息?

**Comment by hyao1** [comment]
很高兴您对这个工作感兴趣，您方便的话可以把复现的代码开源出来，相互学习一下。

feature_copy只是是用来更新类中心特征的，因为我们不需要更新类中心的特征，所以设置不回传梯度来更新类中心特征。计算损失的时候会用到带梯度的特征，可以回传损失更新模型。

这里的归一化并没有什么作用，因为我们只是在class table里选一个最大的值，归一化之后不会影响相似度值的排序。应该我在整理代码的时候忘记删掉了。

**Comment by jiangguisyp** [comment]
> 感谢您的工作, 我最近在我的工作里复现ECCLoss, 发现在细粒度小数据集上的确提升明显, 不过当我和这里的ECCLoss实现进行对比的时候, 我发现有一些细微的差异, 有一些疑惑的地方, 希望能帮忙解答
> 
> [ECC/utils.py](https://github.com/hyao1/ECC/blob/6a214246361b1b039ddcfe0041bc7e0117f80d4e/utils.py#L29)
> 
> Line 29 in [6a21424](/hyao1/ECC/commit/6a214246361b1b039ddcfe0041bc7e0117f80d4e)
> 
>  feature_copy = feature.clone().detach() 
> 
> 这里detach出来之后, `feature_copy`和`feature_table`都没计算梯度, 那`feature_intra_loss` 是否会回传梯度?
> [ECC/utils.py](https://github.com/hyao1/ECC/blob/6a214246361b1b039ddcfe0041bc7e0117f80d4e/utils.py#L43)
> 
> Line 43 in [6a21424](/hyao1/ECC/commit/6a214246361b1b039ddcfe0041bc7e0117f80d4e)
> 
>  class_table = (class_table - torch.min(class_table)) / (torch.max(class_table) - torch.min(class_table)) 
> 
> 这里为什么要做归一化, 我看论文中没有提及, 也可能是我遗漏了什么信息?

感谢大佬的工作，最近我在阅读大佬的论文跟代码，跟上面同学有一样的疑问：
https://github.com/hyao1/ECC/blob/6a214246361b1b039ddcfe0041bc7e0117f80d4e/utils.py#L48
        feature_intra_loss = (torch.cosine_similarity(feature_copy, similar_class_feature[targets], dim=1) * similar_target_value).sum()

这行代码中，feature_copy是detach的，similar_class_feature是来自self.feature_table，requires_grad=False也是没有梯度的，因此这个loss只是一个常量，没有任何回传梯度，并不会对模型优化产生影响，这个是笔误么？
